File: moqt/track_writer.go
package moqt

import (
	"context"
	"errors"
	"fmt"
	"sync"
	"sync/atomic"

	"github.com/okdaichi/gomoqt/moqt/internal/message"
	"github.com/okdaichi/gomoqt/transport"
)

type groupWriterManager struct {
	mu           sync.Mutex
	activeGroups map[*GroupWriter]struct{}

	closed bool
}

func newGroupWriterManager() *groupWriterManager {
	return &groupWriterManager{
		activeGroups: make(map[*GroupWriter]struct{}),
	}
}

func (m *groupWriterManager) addGroup(group *GroupWriter) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.closed {
		return
	}
	m.activeGroups[group] = struct{}{}
}

func (m *groupWriterManager) removeGroup(group *GroupWriter) {
	m.mu.Lock()
	defer m.mu.Unlock()

	delete(m.activeGroups, group)
}

func (m *groupWriterManager) countGroups() int {
	m.mu.Lock()
	defer m.mu.Unlock()
	return len(m.activeGroups)
}

func (m *groupWriterManager) close() {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.closed = true
	m.activeGroups = nil
}

func newTrackWriter(
	broadcastPath BroadcastPath,
	trackName TrackName,
	subscribeStream *receiveSubscribeStream,
	openUniStreamFunc func() (transport.SendStream, error),
	onCloseTrackFunc func(),
) *TrackWriter {
	streamCtx := subscribeStream.stream.Context()

	track := &TrackWriter{
		BroadcastPath:     broadcastPath,
		TrackName:         trackName,
		subscribeStream:   subscribeStream,
		groupManager:      newGroupWriterManager(),
		openUniStreamFunc: openUniStreamFunc,
		onCloseTrackFunc:  onCloseTrackFunc,
		ctx:               context.WithValue(streamCtx, biStreamTypeCtxKey, message.StreamTypeSubscribe),
	}

	return track
}

// TrackWriter writes groups for a published track.
// It manages the lifecycle of active groups for that track.
// The TrackWriter provides output-side methods to send track data.
type TrackWriter struct {
	BroadcastPath BroadcastPath
	TrackName     TrackName

	subscribeStream *receiveSubscribeStream

	groupManager *groupWriterManager

	mu sync.RWMutex

	// groupSequence is atomically incremented for each OpenGroup call
	groupSequence atomic.Uint64

	openUniStreamFunc func() (transport.SendStream, error)

	onCloseTrackFunc func()

	ctx context.Context
}

// Close stops publishing and cancels active groups.
func (w *TrackWriter) Close() error {
	// Take the write lock to ensure Close is exclusive with OpenGroup calls.
	// This prevents OpenGroup from running concurrently with Close and
	// provides a deterministic semantics: either the OpenGroup completes
	// entirely before Close proceeds, or Close waits for OpenGroup to finish.
	w.mu.Lock()
	defer w.mu.Unlock()

	if w.groupManager != nil {
		groupManager := w.groupManager
		w.groupManager = nil

		activeGroups := groupManager.activeGroups

		groupManager.close()

		for g := range activeGroups {
			_ = g.Close()
		}
	}

	if w.onCloseTrackFunc != nil {
		w.onCloseTrackFunc()
		w.onCloseTrackFunc = nil
	}

	if w.subscribeStream == nil {
		return nil
	}

	return w.subscribeStream.close()
}

// CloseWithError stops publishing due to an error and cancels active groups.
func (w *TrackWriter) CloseWithError(code SubscribeErrorCode) {
	// Ensure CloseWithError is exclusive with OpenGroup.
	w.mu.Lock()
	defer w.mu.Unlock()

	if w.groupManager != nil {
		groupManager := w.groupManager
		w.groupManager = nil

		activeGroups := groupManager.activeGroups

		groupManager.close()

		for g := range activeGroups {
			g.CancelWrite(PublishAbortedErrorCode)
		}
	}

	if w.onCloseTrackFunc != nil {
		w.onCloseTrackFunc()
		w.onCloseTrackFunc = nil
	}

	if w.subscribeStream != nil {
		w.subscribeStream.closeWithError(code)
	}
}

// OpenGroup opens a new group with an automatically incremented sequence number
// and returns a GroupWriter to write frames into it.
// The sequence starts at 1 and increments by 1 for each call.
func (w *TrackWriter) OpenGroup() (*GroupWriter, error) {
	// Atomically increment and get the next sequence
	seq := GroupSequence(w.groupSequence.Add(1))
	return w.openGroupWithSequence(seq)
}

// OpenGroupAt opens a new group with the specified sequence number.
// It advances the internal next-sequence counter to at least seq+1 so that
// subsequent OpenGroup calls will not produce a duplicate sequence.
func (w *TrackWriter) OpenGroupAt(seq GroupSequence) (*GroupWriter, error) {
	// Advance the internal counter to avoid collisions with subsequent
	// OpenGroup calls. CAS loop ensures correctness under concurrency.
	for {
		cur := w.groupSequence.Load()
		next := max(cur, uint64(seq)+1)
		if next == cur {
			break
		}
		if w.groupSequence.CompareAndSwap(cur, next) {
			break
		}
	}
	return w.openGroupWithSequence(seq)
}

// SkipGroups skips the next n group sequences without opening them.
// This is useful when you need to intentionally create gaps in the sequence,
// for example, when dropping groups due to packet loss or priority decisions.
func (w *TrackWriter) SkipGroups(n uint64) {
	w.groupSequence.Add(n)
}

func (w *TrackWriter) Context() context.Context {
	return w.ctx
}

func (w *TrackWriter) WriteInfo(info PublishInfo) error {
	return w.subscribeStream.writeInfo(info)
}

// DropGroups sends a SUBSCRIBE_DROP message for an explicit inclusive range.
// The range is expressed in absolute group sequence numbers.
func (w *TrackWriter) DropGroups(drop SubscribeDrop) error {
	if drop.StartGroup == MinGroupSequence || drop.EndGroup == MinGroupSequence {
		return fmt.Errorf("invalid drop range: start=%d end=%d", drop.StartGroup, drop.EndGroup)
	}
	if drop.StartGroup > drop.EndGroup {
		return fmt.Errorf("invalid drop range: start=%d end=%d", drop.StartGroup, drop.EndGroup)
	}

	w.mu.RLock()
	defer w.mu.RUnlock()

	if w.Context().Err() != nil {
		return Cause(w.Context())
	}

	if w.subscribeStream == nil {
		return fmt.Errorf("track writer is closed")
	}

	return w.subscribeStream.writeDrop(drop)
}

// DropNextGroups skips the next n groups and emits a SUBSCRIBE_DROP for the
// skipped inclusive range.
func (w *TrackWriter) DropNextGroups(n uint64, code SubscribeErrorCode) error {
	if n == 0 {
		return nil
	}

	for {
		cur := w.groupSequence.Load()
		start := GroupSequence(cur + 1)
		end := GroupSequence(cur + n)

		if end < start {
			return fmt.Errorf("drop range overflow: start=%d end=%d", start, end)
		}

		if w.groupSequence.CompareAndSwap(cur, uint64(end)) {
			return w.DropGroups(SubscribeDrop{
				StartGroup: start,
				EndGroup:   end,
				ErrorCode:  code,
			})
		}
	}
}

func (w *TrackWriter) TrackConfig() *SubscribeConfig {
	if w.subscribeStream == nil {
		return &SubscribeConfig{}
	}

	return w.subscribeStream.TrackConfig()
}

func (w *TrackWriter) Updated() <-chan struct{} {
	return w.subscribeStream.Updated()
}

// openGroupWithSequence is the internal implementation for opening a group with a specific sequence.
func (w *TrackWriter) openGroupWithSequence(seq GroupSequence) (*GroupWriter, error) {
	// Avoid accessing s.ctx directly; it can be nil if the receiveSubscribeStream
	// has been cleared during Close(). Instead, capture the receiveSubscribeStream
	// under lock and validate its context below.

	// Prevent opening a new group if the track has been closed. Capture
	// receiveSubscribeStream under lock so it cannot be set to nil by Close()
	// Acquire a shared lock so multiple OpenGroup calls can proceed
	// concurrently while ensuring Close waits for them to finish.
	w.mu.RLock()
	defer w.mu.RUnlock()

	// Check the context on the captured receiveSubscribeStream instead of s.ctx
	// to avoid nil deref if the embedded field has been cleared by Close().
	if w.Context().Err() != nil {
		return nil, Cause(w.Context())
	}

	// Write the INFO message to the receive subscribe stream.
	err := w.WriteInfo(PublishInfo{
		StartGroup: seq,
	})
	if err != nil {
		return nil, err
	}

	stream, err := w.openUniStreamFunc()
	if err != nil {
		if appErr, ok := errors.AsType[*transport.ApplicationError](err); ok {
			sessErr := &SessionError{
				ApplicationError: appErr,
			}
			return nil, sessErr
		}
		return nil, err
	}

	err = message.StreamTypeGroup.Encode(stream)
	if err != nil {
		var strErr *transport.StreamError
		if errors.As(err, &strErr) {
			return nil, &GroupError{StreamError: strErr}
		}

		strErrCode := transport.StreamErrorCode(InternalGroupErrorCode)
		stream.CancelWrite(strErrCode)

		return nil, err
	}

	err = message.GroupMessage{
		SubscribeID:   uint64(w.subscribeStream.subscribeID),
		GroupSequence: uint64(seq),
	}.Encode(stream)
	if err != nil {
		var strErr *transport.StreamError
		if errors.As(err, &strErr) {
			return nil, &GroupError{StreamError: strErr}
		}

		strErrCode := transport.StreamErrorCode(InternalGroupErrorCode)
		stream.CancelWrite(strErrCode)

		return nil, err
	}

	return newGroupWriter(stream, seq, w.groupManager), nil
}

File: moq-web/src/track_writer.ts
import { GroupWriter } from "./group_stream.ts";
import type { Info } from "./info.ts";
import { type Context, ContextCancelledError, toAbortSignal } from "@okdaichi/golikejs/context";
import type { ReceiveSubscribeStream, SubscribeDrop, TrackConfig } from "./subscribe_stream.ts";
import type { SendStream } from "./internal/webtransport/mod.ts";
import { UniStreamTypes } from "./stream_type.ts";
import { GroupMessage, writeVarint } from "./internal/message/mod.ts";
import type { BroadcastPath } from "./broadcast_path.ts";
import type { SubscribeErrorCode } from "./error.ts";
import { GroupErrorCode } from "./error.ts";

/**
 * Publisher-side handle for writing groups to a subscribed track.
 *
 * Created by the {@link TrackMux} when an incoming subscription is matched.
 * Use {@link openGroup} to start writing frames.
 */
export class TrackWriter {
	/** The broadcast path this track belongs to. */
	broadcastPath: BroadcastPath;
	/** The track name within the broadcast. */
	trackName: string;
	#subscribeStream: ReceiveSubscribeStream;
	#openUniStreamFunc: (signal?: AbortSignal) => Promise<
		[SendStream, undefined] | [undefined, Error]
	>;
	#groups: GroupWriter[] = [];
	#groupCount: number = 0;

	constructor(
		broadcastPath: BroadcastPath,
		trackName: string,
		subscribeStream: ReceiveSubscribeStream,
		openUniStreamFunc: (signal?: AbortSignal) => Promise<
			[SendStream, undefined] | [undefined, Error]
		>,
	) {
		this.broadcastPath = broadcastPath;
		this.trackName = trackName;
		this.#subscribeStream = subscribeStream;
		this.#openUniStreamFunc = openUniStreamFunc;
	}

	get context(): Context {
		return this.#subscribeStream.context;
	}

	get subscribeId(): number {
		return this.#subscribeStream.subscribeId;
	}

	get config(): TrackConfig {
		return this.#subscribeStream.trackConfig;
	}

	/**
	 * Open a new group with an auto-incremented sequence number.
	 *
	 * The group opens on a fresh unidirectional stream; when the peer's
	 * concurrent-stream limit is reached this blocks until a stream is granted.
	 * Pass `options.signal` to cancel or time out the open. If aborted, no group
	 * stream remains open.
	 * @param options - Optional cancellation `signal`.
	 * @returns A {@link GroupWriter} for writing frames, or an error.
	 */
	async openGroup(
		options?: { signal?: AbortSignal },
	): Promise<[GroupWriter, undefined] | [undefined, Error]> {
		this.#groupCount++;
		return this.#openGroupWithSequence(this.#groupCount, options?.signal);
	}

	/**
	 * Open a new group at a specific sequence number.
	 * @param seq - The group sequence number.
	 * @param options - Optional cancellation `signal`.
	 */
	async openGroupAt(
		seq: number,
		options?: { signal?: AbortSignal },
	): Promise<[GroupWriter, undefined] | [undefined, Error]> {
		if (seq > this.#groupCount) {
			this.#groupCount = seq;
		}
		return this.#openGroupWithSequence(seq, options?.signal);
	}

	/** Advance the group counter by `n` without opening groups. */
	skipGroups(n: number): void {
		this.#groupCount += n;
	}

	/**
	 * Notify the subscriber that groups in the given range were dropped.
	 * @param drop - Range and error code of dropped groups.
	 */
	async dropGroups(drop: SubscribeDrop): Promise<Error | undefined> {
		if (drop.startGroup === 0 || drop.endGroup === 0) {
			return new Error(`invalid drop range: start=${drop.startGroup} end=${drop.endGroup}`);
		}
		if (drop.startGroup > drop.endGroup) {
			return new Error(`invalid drop range: start=${drop.startGroup} end=${drop.endGroup}`);
		}
		return this.#subscribeStream.writeDrop(drop);
	}

	async dropNextGroups(n: number, code: SubscribeErrorCode): Promise<Error | undefined> {
		if (n === 0) {
			return undefined;
		}
		const start = this.#groupCount + 1;
		const end = this.#groupCount + n;
		this.#groupCount = end;

		return this.dropGroups({
			startGroup: start,
			endGroup: end,
			errorCode: code,
		});
	}

	async updated(): Promise<void> {
		return this.#subscribeStream.updated();
	}

	async #openGroupWithSequence(
		seq: number,
		signal?: AbortSignal,
	): Promise<[GroupWriter, undefined] | [undefined, Error]> {
		// Mirror Go openGroupWithSequence (track_writer.go:298): if the track
		// context is already done, fail fast without opening a stream.
		if (this.context.err()) {
			return [undefined, this.context.err()!];
		}

		// ensureInfo is not raced against the signal (Go passes no ctx here).
		let err: Error | undefined;
		err = await this.#subscribeStream.ensureInfo();
		if (err) {
			return [undefined, err];
		}

		// Effective cancellation: the track context OR the caller's signal.
		// AbortSignal.any composes them; with no caller signal the track context
		// alone bounds the open (Go's `ctx.Done()==nil → w.Context()`).
		const trackSignal = toAbortSignal(this.context);
		const effective: AbortSignal = signal
			? AbortSignal.any([trackSignal, signal])
			: trackSignal;

		// §1 Abort before starting: do not open a stream.
		if (effective.aborted) {
			const reason = (effective as unknown as { reason?: unknown }).reason;
			return [
				undefined,
				reason instanceof Error ? reason : new ContextCancelledError(),
			];
		}

		// Thread the effective signal down to the stream-open, mirroring Go's
		// openUniStreamFunc(ctx) (track_writer.go:314).
		let writer: SendStream | undefined;
		[writer, err] = await this.#openUniStreamFunc(effective);
		if (err) {
			return [undefined, err];
		}

		// §3 A stream is now allocated. Any failure before the group header is
		// fully written must cancel the stream so it is never left half-open.
		// (Mirrors Go stream.CancelWrite on encode error; previously leaked.)
		[, err] = await writeVarint(writer!, UniStreamTypes.GroupStreamType);
		if (err) {
			await writer!.cancel(GroupErrorCode.InternalError).catch(() => {});
			return [undefined, new Error(`Failed to write stream type: ${err}`)];
		}

		const msg = new GroupMessage({
			subscribeId: this.subscribeId,
			sequence: seq,
		});
		err = await msg.encode(writer!);
		if (err) {
			await writer!.cancel(GroupErrorCode.InternalError).catch(() => {});
			return [undefined, new Error("Failed to create group message")];
		}

		const group = new GroupWriter(this.context, writer!, msg);

		this.#groups.push(group);

		return [group, undefined];
	}

	/**
	 * Write publisher {@link Info} to the subscriber.
	 * @param info - Track information to send.
	 */
	async writeInfo(info: Info): Promise<Error | undefined> {
		const err = await this.#subscribeStream.writeInfo(info);
		if (err) {
			return err;
		}

		return undefined;
	}

	async closeWithError(code: SubscribeErrorCode): Promise<void> {
		// Cancel all groups with the error first
		await Promise.allSettled(this.#groups.map(
			(group) => group.cancel(GroupErrorCode.PublishAborted),
		));

		// Then close the subscribe stream with the error
		await this.#subscribeStream.closeWithError(code);
	}

	async close(): Promise<void> {
		await Promise.allSettled(this.#groups.map(
			(group) => group.close(),
		));

		await this.#subscribeStream.close();
	}
}
